### ISSUE TITLE
injectDecodedDataToFilterChain and injectEncodedDataToFilterChain cannot be used after header StopIteration

### ISSUE THREAD

**Issue by ravenblackx** [envoyproxy/envoy#19891]
## Description:
When creating a filter that performs internal buffering and delivers the data asynchronously, it is currently impossible to both postpone the header *and* stream the content.

### Option 1: postpone the header, use addDecodedData / addEncodedData

This appears to work, but it actually puts the entirety of the data into a memory buffer until you call `continueEncoding`/`continueDecoding` - if you honor the buffer limits and they are not big enough, this results in a deadlock.

### Option 2: postpone the header, use addDecodedData / addEncodedData, and call continueDecoding after each insertion

This appears to work, but the first time you call continueDecoding after the end of the *input* stream has been reached (not the end of the data you're adding), the connection terminates and whatever data you were intending to add is lost.

### Option 3: postpone the header, use injectDecodedDataToFilterChain/injectEncodedDataToFilterChain

This throws an assertion because the inject.* functions require that the filter be in a buffering state, which it's not until after headers are sent.

### Option 4: explicitly send the headers before using injectDecodedDataToFilterChain etc.

There is no way to explicitly send the headers, there is only `continueDecoding`, which, if the end of the incoming stream has been seen, terminates the stream.

## Solution:

It turns out there's a [TODO](https://github.com/envoyproxy/envoy/blob/c5dca643609ed4b7b9da5d7f1078547b2c36d614/source/common/http/filter_manager.cc#L1588) in the code about this. I am constructing a pull request to resolve that TODO.

**Comment by htuch** [comment]
CC @mattklein123
